### ISSUE TITLE
Notebook api high level feedback 

### ISSUE THREAD

**Issue by mjbvz** [microsoft/vscode#121329]
I've collected some thoughts on the sections of the notebook that we'd like to finalize first. This includes:

- Serializer
- Kernel
- NotebookDocument

This feedback focuses more on the API design itself rather than how the api actually works since I know there's been a huge amount of discussion around the notebook API already 

## General feedback

- Add `readonly` to all properties on event objects and all "options" style objects

- There are a few TODOs in the code already. I skipped over much of the code that was marked with TODO

## NotebookSerializer

- `dataToNotebook` -> `deserializeNotebook` (we already have `deserializeWebviewPanel`)

- `notebookToData` -> `serializeNotebook`

- Add cancellation tokens to both methods 

## NotebookData

- Should this be an interface instead?

    As far as I can tell, it is only used to pass around the notebook data. Is the object stored by VS Code?

- Same with `NotebookCellData`. Should it also be an interface?

    If you subclass `NotebookCellData`, are you ever later passed back the same object?

## NotebookDocumentMetadata

- Maybe this should be an interface too?
 
    The main helper that it brings is `.with`, but it may be better to express that use JS spreads:

    ```js
    const newMetadata = { trusted: true, ...oldMetadata };
    ```

## NotebookKernel

- Rename to `NotebookController` to make propose more clear

- Better name for `NotebookKernelOptions`? You're basically specifying the implementation of the kernel here, so `options` feels a bit odd

    I wonder if this object should be split up too? Something like:

    ```ts
    createNotebookKernel(id, selector, { ...executeHandler })
    ```

- `NotebookKernelOptions` properties should all be readonly

- Extract the complex types from `NotebookKernelOptions` to their own type alias?

    `executeHandler` for example.

- Should `NotebookKernel2` expose properties like `executeHandler`? 

    These are used internally but should an extension ever call them directly?

- Can a kernel ever be disposed of by VS Code or does the extension fully manage its lifecycle?

## NotebookCellExecutionTask

- Can we encode that `start` should always be called? (I think it should at least)

    Maybe something closer to our edit builder apis?

    ```ts
    interface NotebookCellExecutionTask {
        ...

        execute(
            context: NotebookCellExecuteStartContext,
            run: (executor: NotebookCellExecutor) => Promise<NotebookCellExecuteResult>
        ): void;
    }

    interface NotebookCellExecutor {
        appendOutput(out: NotebookCellOutput | NotebookCellOutput[], cellIndex?: number): Thenable<void>;
        ...
    }
    ```

    example usage:

    ```ts
    task.execute({}, async (builder, token) => {
        await builder.appendOutput(...)
        ...
        return { success: true };
    })
    ```

- Do we need both `createNotebookCellExecutionTask` and `NotebookKernel2` createNotebookCellExecutionTask?

- Should these top level functions take cancellation tokens?

- Instead of `append` and `replace`, could we just have a single `splice` method?

- When talking about cell indexes, do these reflect the indexes when the notebook first started executing or after previous edits have already been made?

- Do I need to await the result of the edit operations before calling `end`?

## NotebookDocument

- remove `fileName` since it's already deprecated

- Do we want to allow extensions to call `save`? I thought we discouraged this for `TextDocument`

## NotebookCell

- Why do we need a link back to the document?

- `NotebookCellExecutionSummary` properties should all be `readonly`

### NotebookCellMetadata

- Same as `NotebookDocumentMetadata`, should this be an interface instead?

**Comment by rebornix** [comment]
The current shape of `NotebookCellMetadata`. The `with` API helps with doing incremental update of predefined metadata but doesn't help too much with custom metadata. 

```ts
export class NotebookCellMetadata {
	/**
	* Whether a code cell's editor is collapsed
	*/
	readonly inputCollapsed?: boolean;

	/**
	* Whether a code cell's outputs are collapsed
	*/
	readonly outputCollapsed?: boolean;

	/**
	* Additional attributes of a cell metadata.
	*/
	readonly [key: string]: any;

	constructor(inputCollapsed?: boolean, outputCollapsed?: boolean);

	with(change: { inputCollapsed?: boolean | null, outputCollapsed?: boolean | null, [key: string]: any }): NotebookCellMetadata;
}
```

**Comment by jrieken** [comment]
> For NotebookData/NotebookCellData/NotebookDocumentMetadata/NotebookCellMetadata, interfaces with non-readonly properties is the most convenient to use.

I think we never spelled that our in API guidelines but for me classes are always the preferable solution when extension are expected to give us data. That is mostly due to the convenience of a constructor which helps you create a "correct" object. I think with providers we are consistent here, like `CompletionItem` or `Hover`

And also we should not forget that TypeScript largely ignores the differences between classes and interfaces, e.g TS these two variants are equal in TS's eyes and when implementing the API we cannot use `instanceof`-checks

```ts
  // interfaces
  const result: vscode.NotebookData = {
      metadata: new vscode.NotebookDocumentMetadata(),
      cells: []
  }

  // classes, ctor sets default metadata
  const result2 = new vscode.NotebookData([]);

```

**Comment by mjbvz** [comment]
For `NotebookController`, @rebornix and I talked about splitting `executionHandler` to a new interface:

```ts
	export interface NotebookControllerExecutionHandler {
		executeNotebookCells(cells: NotebookCell[], controller: NotebookController): void;
		interruptNotebookExecution?(notebook: NotebookDocument): void;
	}

	export namespace notebook {
		export function createNotebookController(
			options: NotebookControllerOptions,
			executionHandler: NotebookControllerExecutionHandler
		): NotebookController;
	}
```

**Comment by rebornix** [comment]
What I ended up with adopting the execution api in Kusto notebook is creating a `kernel` class, which creates a controller internally and hold a reference to it:

```ts
export class Kernel {
    controller: NotebookController;
    constructor() {
        this.controller = notebook.createNotebookController({
            id: 'kusto',
            label: 'Kusto',
            description: 'Execute Kusto Queries',
            selector: { viewType: 'kusto-notebook' },
            supportedLanguages: ['kusto'],
            executeHandler: this.execute.bind(this)
        });
        this.controller.interruptHandler = this.interrupt;
        this.controller.isPreferred = true;
    }

    interrupt() {}
    execute() {}
}
```

Conceptually the `Kernel` is an execution delegate for the `NotebookController` so with Matt's proposal above, the code can be simplied as 

```ts
export class Kernel implements NotebookControllerExecutionHandler {
    controller: NotebookController;
    constructor() {
        this.controller = notebook.createNotebookController({
            id: 'kusto',
            label: 'Kusto',
            description: 'Execute Kusto Queries',
            selector: { viewType: 'kusto-notebook' },
            supportedLanguages: ['kusto'],
        });
        this.controller.executionHandler = this;
    }

    interrupt() {}
    execute() {}
}
```

IMHO This now has a clear cut of responsibilities of the Kernel and Controller objects: `Kernel` is handling execution/interrupt, and `Controller` is for manipulating the UI execution state.

**Comment by jrieken** [comment]
> IMHO This now has a clear cut of responsibilities of the Kernel and Controller objects: Kernel is handling execution/interrupt, and Controller is for manipulating the UI execution state.

I think both variants are very, very similar. The difference boils down to being able to assign an object which owns execute and interrupt or being able to assign both functions directly. Overall, I don't see the cut (between kernel and controller) as so clear because you always need the controller to create execution tasks. For me there is just the controller (or however you wanna name it) and that does all the things.

**Comment by rebornix** [comment]
> I think both variants are very, very similar

yes they are really close, in JavaScript world we can even call them identical and won't change how one will organize their code. But delegating all execution related responsibility (execution callback and optional interrupt handler) to an object fits my mental model better (the `NotebookControllerExecutionHandler` proposal).

**Comment by IanMatthewHuff** [comment]
On the Jupyter side we ended up with something similar to @rebornix in our code. Our class wrapping the NotebookController constructor we're actually just calling the NotebookController as well. Our concept of kernel has a lower level meaning for us and we liked the name NotebookController to represent this VS Code Kernel Selection => Execution Handler concept that sits above our kernels. Our main transitions were from our old kernelProvider to a new notebookControllerManager and from something that we called KernelWithMetadata to our own NotebookController class. 

We have a PR up that should at least keep things from breaking tomorrow, but for sure we won't have our widgets support yet and our tests need fixing as well, just not enough time to update the test bits that were using the old KernelProvider machinery yet.

PR here if you are interested in peeking: https://github.com/microsoft/vscode-jupyter/pull/5565/files

Some comments on the API here. 

- Previously we used to provide a full array of kernels, and mark the initial one that we wanted selected via is preferred. isPreferred exists on the new NotebookController, but I'm not sure if this really works as these are created one at a time via createNotebookController. Marking something isPreferred just moves it to the top of the selection list, the kernel that will be initially selected for a document always seems to just be the first thing that you pass to createNotebookController. Easy enough to get around by just moving our preferred item to the top of the list that we create NotebookControllers with. But if we are just sorting the list and then calling createNotebookController in order then I'm not sure why we need isPreferred anymore? 
- In the old NotebookKernel we used to explicitly not set supported languages, it was optional, and if you didn't specify this you would support all vscode languages. Supported languages is now required for a cell to execute, and can't be left empty. We support anything that you can launch a kernel for in Jupyter so we'd actually like to just support everything again, working around this for now just by passing in a big known list, but we might be ruling out some oddball kernels that we could support by doing this.
- The onDidChangeNotebookAssociation event was a bit funny to adapt. We still needed something like the old onDidChangeActiveNotebookKernel. So we mapped the onDidChangeNotebookAssociation event to include the controller as well as the document a level up. And then used our controllerManager to create a general onNotebookControllerSelected event that other classes could hook up to see then a document's controller was changed. That worked, but onDidChangeNotebookAssociation doesn't seem to fire for the selection of the initial kernel, only when you open the picker and pick a new one, so we had to tweak it a bit for that as well. 

I think @DonJayamanne might have some feedback on the communications part of this as well as he's looking at that for our widget support.

**Comment by DonJayamanne** [comment]
@rebornix @jrieken 
We're currently reviewing how notebook trust is implemented in Jupyter. Today when a notebook isn't trusted we prevent execution, editing, etc. In fact we might just want the ability to just hide output when a notebook is not trusted.
If a user executes a cell, then the new output of that cell would be deemed as trusted. The rest would still be untrusted, then if the user runs another cell, that (new) out putt will be displayed.

I.e. we just need the ability to hide output of cells, trust output at a cell level.
That's how classic Jupyter notebooks implements trust today.
Again we're reviewing this in Jupyter extension.

Please feel free to move this discussion to a separate issue, not sure where it belongs.

**Comment by jrieken** [comment]
> Marking something isPreferred just moves it to the top of the selection list, the kernel that will be initially selected for a document always seems to just be the first thing that you pass to createNotebookController. Easy enough to get around by just moving our preferred item to the top of the list that we create NotebookControllers with. But if we are just sorting the list and then calling createNotebookController in order then I'm not sure why we need isPreferred anymore?

Well, in theory there can be a controller from another extension targeting the same notebook. For that case isPreferred is useful. Tho, it is true that `isPreferred` isn't so important anymore. The [sorting goes by](https://github.com/microsoft/vscode/blob/0c723c1bf37b02c15cad029eb3c047c994df5915/src/vs/workbench/contrib/notebook/browser/notebookKernelServiceImpl.ts#L137-L150): isPreferred, score (*), and creation time. (* the score is how well the notebook selector matched on the notebook)

> In the old NotebookKernel we used to explicitly not set supported languages, it was optional, and if you didn't specify this you would support all vscode languages. 

Yeah, it was an arbitrary decision on our side, e.g when there is no supported languages we use 'text/plain' but we can also change the semantics to say "all languages". However, I am surprised that jupyter supports all languages. Is this restricted by the actual jupyter-kernel that is being picked, like my JavaScript-jupyter kernel cannot do python? Anyways, let me change the default and make you life simpler :-)

> We still needed something like the old onDidChangeActiveNotebookKernel. 

We had to remove that because otherwise your controller instance is becoming "public" and any extension would be able to do stuff with it

**Comment by DonJayamanne** [comment]
### 1. Global even for when a NotebookController changes
Today we need to add an event on each notebook controller to determine whether its the active controller or not.
This seems rather clunky, I'd prefer the global events liks `onDidChangeNotebookController({NotebookDocument, NotebookControllerId})`

I believe the reason we don't have a global even is to prevent other extension authors getting access to the notebook controller. I believe we can address this by passing the `NotebookControllerId`.

Right now we need to do a few things when ever the controller (kernel) changes. This happens in a few places. All of thise needs to change to add handlders to all of the individual controllers to check if it is the active/selected or not. We can also re-route all of these event handlers and have our own global event that tells our code when this happens.

However this seems like a work around for the VSC API & makes our code a little bit more confusing (doesn't like up with the rest of the VSC API & other patterns we seem to have). 
We would prefer if VSC instead had an event `onDidChangeNotebookController`.

* VSCode could fire an event with the notebookDocument & the id of the controller.
* Rather than extension authors providing an Id, VS Code could generate an ID.
	(this would get around the issue of conflicting ids)

### 2. Currently Active Notebook Controller

I could be wrong here, but based on my understanding this is what happens today:
* User opens a notebook a.ipynb
* Extension creates a notebook controller A, this is now the active notebook controller
* Extension creates a notebook controller B, C (this is preferred), D
* C is preferred but not the active controller.

I prefer the previous implementation over this.
There's no clear contract that tells us (extension authors) what the active/selected controller is when a notebook is opened.
Picking the first controller as the active doesn't seem right in the push model.

I'd prefer the following:
**Scenario 1**
* User opens a notebook a.ipynb
* Extension creates a notebook controller A
* No active notebook controllers/kernels.
* User sees a message in status bar `Select kernel`, when user attempts to run a cell, then they get the picker to select a kernel (this is how it worked previously with provider model)

**Scenario 2**
* User opens a notebook a.ipynb
* Extension creates a notebook controller A
* User sees a message in status bar `Select kernel`
* After a few seconds the extension pushes more controllers...
* Extension creates a notebook controller B, C (this is preferred), D
* C is active controller.
* The status bar will respect the `isPreferred` setting.

FYI - Such a model is crucial for us. Else picking the first controller as the active controller could break notebooks (e.g. .NET extension registering a controller before Jupyter extension could result in that getting picked over Python kernel for a Python Notebook or the like).

TLDR - I'd prefer to have an explicit contract that tells VS Code what should be used as the kernel (isPreferred & if not provided, then there's none).
Similarly if such a kernel is selected by VS Code, extension authors need to know about this (previously the event would be fired).
Though the down side was we didn't know whether the event was fired as a result of user changing the kernel or vs code automatically picking one.

**Comment by DonJayamanne** [comment]
> However, I am surprised that jupyter supports all languages. Is this restricted by the actual jupyter-kernel that is being picked, like my JavaScript-jupyter kernel cannot do python? Anyways, let me change the default and make you life simpler :-)

Replied here https://github.com/microsoft/vscode-jupyter/issues/5588#issue-864040362
Sorry for the separate issue (this is a feature of Jupyter, magics).
Its possible JavaScript can run Python code, but spinning up a Python process (in such a case user can enter Python code for a notebook using a JS kernel).
This is exactly how the SOS kernel & others (like our .NET interacgive kernel) work

Unfortunately the Jupyter Kernel Spec has no way to define the list of supported sub lanaguges either.
Simply not possible, with magics its possible to add support for new langauges when the kernel is the same

**Comment by DonJayamanne** [comment]
### NotebookController.onDidReceiveMessage

If we have a notebook and its active kernel is A.
Now we get messages from `onDidReceiveMessage` on this instance `A`.
Assume user changes the kernel, hence controller also changes to B.
Does this mean messages will now come from `B.onDidReceiveMessage`.

It feels weird as we need to switch the handlers from A to B to get messages.
Or we could add event handlers to all of these NotebeookControllers.
Either that doesn't feel right. Also concerned that messages could be lost while the switch is taking place (basically a lot of moving parts).

I think it would be simpler to have postMessage & onDidReceiveMessage on the NotebookEditor itself. I don't think we need a controller for this (basically decouple this).
Or is the fear that other extensions could listen in on these messages & could open a security hole.

**Comment by DonJayamanne** [comment]
# NotebookController.interruptHandler

Please can we pass the `NotebookDocument` to this method.
This way we can use the same notebookcontroller for all documents.
Today we create multiple, one for each Document.
I'd prefer to create a single set of controllers.

@IanMatthewHuff @jrieken  /cc

**Comment by DonJayamanne** [comment]
### NotebookController.preloads

Question: Should this be a `Uri.file` or a Uri converted into `asWebviewUri`

**Comment by DonJayamanne** [comment]
> 2. Currently Active Notebook Controller

Further to this, the `isPreferred` only works per document.
Hence not sure the `isPreferred` works for Jupyter.

I'd prefer not to have n controllers, one per each document.
I'd prefer to have a method `setPreferredController(NotebookDocument, NotebookController)`

**Comment by mjbvz** [comment]
Sorry for this super nit picky point:

- The `notebookOutputRenderer` and `notebookProvider` contribution points both be plural, i.e. `notebookOutputRenderers` and `notebookProvider`
  
   I believe the rest of our contribution points use plural names and these points both take arrays of contributions

**Comment by jrieken** [comment]
![pause](https://media.giphy.com/media/9Dgf7QW41rueymSlp0/giphy.gif)

Sorry, but can we please create **separate issues** per feedback item. This is becoming unmanageable and we cannot reply to and discuss 10 different topics in one issue

**Comment by jrieken** [comment]
To make this easier to track and discuss, I shelling out some comment to new or existing issues

* package.json-work ➡️ https://github.com/microsoft/vscode/issues/121819
* active/selected controller ➡️  https://github.com/microsoft/vscode/issues/121902

**Comment by DonJayamanne** [comment]
> Sorry, but can we please create separate issues per feedback item.

I followed what others did here, this seems to be a general place for notebook api feedback.

**Comment by mjbvz** [comment]
Yes sorry if this was not clear @DonJayamanne: this issue was only meant to collect my notes on the current api while we get it ready for finalization. It was not supposed to be a general api discussion. Please open new issues if you have any feedback on the current api

I think most of my points have been addressed or split into new issues, so I'm going to close this. We'll continue reviewing the api in May
